Add LangChain optimization guide

TomeHirata opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/TomeHirata/mlflow/pull/18836?quickstart=1)

#### Install mlflow from this PR

```
# mlflow
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/18836/merge
# mlflow-skinny
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/18836/merge#subdirectory=libs/skinny
```

For Databricks, use the following command:

```
%sh curl -LsSf https://raw.githubusercontent.com/mlflow/mlflow/HEAD/dev/install-skinny.sh | sh -s pull/18836/merge
```

</p>
</details>

### Related Issues/PRs

n/a

### What changes are proposed in this pull request?

Add LangChain optimization guide

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

<!-- Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change. -->

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/prompts`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/build`: Build and test infrastructure for MLflow
- [x] `area/docs`: MLflow documentation pages

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

<!-- Do not modify or remove any text inside the parentheses. Keep both checkboxes below. -->

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

TomeHirata committed:
Add LangChain optimization guide and update sidebar links for prompt optimization documentation

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

harupy commented on docs/docs/genai/prompt-registry/optimize-prompts/langchain-optimization.mdx:
```diff
@@ -0,0 +1,141 @@
+
+    llm = OpenAI(temperature=0.7)
+
+    chain = LLMChain(llm=llm, prompt=template)
```
does this example work in langchain v1?

B-Step62 replied:
+1, we should not use LLMChain, it has been deprecated since 0.1.

harupy commented on docs/docs/genai/prompt-registry/optimize-prompts/langchain-optimization.mdx:
```diff
@@ -0,0 +1,141 @@
+# Optimizing Prompts for LangChain
+
+<p style={{display: 'flex', justifyContent: 'center', margin: '1em 0'}}>
+  <img src={useBaseUrl("/images/logos/langchain-logo.png")} alt="LangChain Logo" style={{width: 300, objectFit: 'contain'}} />
```
Do we need the logo?

<img width="713" height="447" alt="Screenshot 2025-11-17 at 19 23 33" src="https://github.com/user-attachments/assets/aa369ce3-beec-4613-b48b-12cca2bee0a1" />

harupy commented on docs/docs/genai/prompt-registry/optimize-prompts/langchain-optimization.mdx:
```diff
@@ -0,0 +1,141 @@
+
+## Key Points for LangChain Integration
+
+### Using `to_single_brace_format()`
```
```suggestion
### Template Format Conversion
```

harupy commented on docs/docs/genai/prompt-registry/optimize-prompts/langchain-optimization.mdx:
````diff
@@ -0,0 +1,141 @@
+)
+```
+
+## Key Points for LangChain Integration
````
```suggestion
## Integration Notes
```

TomeHirata commented on docs/docs/genai/prompt-registry/optimize-prompts/langchain-optimization.mdx:
```diff
@@ -0,0 +1,141 @@
+]
+
+# Step 4: Optimize the prompt
+result = mlflow.genai.optimize_prompts(
```
Yes, it's a known issue, I'm talking to the gepa author about the solution

TomeHirata committed:
Update LangChain optimization guide with new prompt registration and agent integration

- Updated installation command to include `litellm`.
- Introduced a new system prompt for agent creation.
- Refactored the prediction function to utilize the new agent and user prompt.
- Removed outdated sections and examples for clarity.

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

TomeHirata committed:
comment

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

harupy commented on docs/docs/genai/prompt-registry/optimize-prompts/langchain-optimization.mdx:
````diff
@@ -0,0 +1,138 @@
+## Prerequisites
+
+```bash
+pip install -U langchain langchain-openai mlflow gepa litellm
````
```suggestion
pip install -U langchain>=1 langchain-openai mlflow gepa litellm
```

in case a user has langchain<1

harupy reviewed (approved):
LGTM once https://github.com/mlflow/mlflow/pull/18836/files#r2541107105 is addressed!
